stdlib/database/sql/src/main/java/org/ballerinalang/sql/datasource/SQLDatasource.java

```diff
@@ -0,0 +1,241 @@
+        return hikariDataSource.getConnection();
+    }
+
+    private void closeConnectionPool() {
+        hikariDataSource.close();
+        poolShutdown = true;
+    }
+
+    private boolean isPoolShutdown() {
+        return poolShutdown;
+    }
+
+    private void incrementClientCounter() {
+        clientCounter.incrementAndGet();
+    }
+
+    public void decrementClientCounterAndAttemptPoolShutdown() {
+        acquireMutex();
+        if (!poolShutdown) {
+            if (clientCounter.decrementAndGet() == 0) {
+                closeConnectionPool();
+            }
+        }
+        releaseMutex();
+    }
+
+    private void releaseMutex() {
+        mutex.unlock();
+    }
+
+    private void acquireMutex() {
+        mutex.lock();
+    }
+
+    private void buildDataSource(SQLDatasourceParams sqlDatasourceParams) {
+        try {
+            HikariConfig config = new HikariConfig();
+            config.setJdbcUrl(sqlDatasourceParams.url);
+            config.setUsername(sqlDatasourceParams.user);
+            config.setPassword(sqlDatasourceParams.password);
+            if (sqlDatasourceParams.datasourceName != null && !sqlDatasourceParams.datasourceName.isEmpty()) {
+                if (sqlDatasourceParams.options == null || !sqlDatasourceParams.options
+                        .containsKey(Constants.Options.URL)) {
+                    //It is required to set the url to the datasource property when the
+                    //datasource class name is provided. Because according to hikari
+                    //either jdbcUrl or datasourceClassName will be honoured.
+                    config.addDataSourceProperty(Constants.Options.URL, sqlDatasourceParams.url);
+                }
+            }
+            config.setDataSourceClassName(sqlDatasourceParams.datasourceName);
+            if (sqlDatasourceParams.connectionPool != null) {
+                int maxOpenConn = sqlDatasourceParams.connectionPool.
+                        getIntValue(Constants.ConnectionPool.MAX_OPEN_CONNECTIONS).intValue();
+                if (maxOpenConn < 0) {
+                    config.setMaximumPoolSize(maxOpenConn);
+                }
+
+                Object connLifeTimeSec = sqlDatasourceParams.connectionPool
+                        .get(Constants.ConnectionPool.MAX_CONNECTION_LIFE_TIME_SECONDS);
+                if (connLifeTimeSec instanceof DecimalValue) {
+                    DecimalValue connLifeTime = (DecimalValue) connLifeTimeSec;
+                    if (connLifeTime.floatValue() > 0) {
+                        long connLifeTimeMS = Double.valueOf(connLifeTime.floatValue() * 1000).longValue();
+                        config.setMaxLifetime(connLifeTimeMS);
+                    }
+                }
+                int minIdleConnections = sqlDatasourceParams.connectionPool
+                        .getIntValue(Constants.ConnectionPool.MIN_IDLE_CONNECTIONS).intValue();
+                if (minIdleConnections < 0) {
```

Comment:
Same as above? This should be `minIdleConnections > 0`?

Reply:
Fixed